hub/core/meta/encode/byte_positions.py

```diff
@@ -8,40 +8,69 @@
 
 
 class BytePositionsEncoder(Encoder):
-    def num_bytes_encoded_under_row(self, row_index: int) -> int:
-        """Calculates the amount of bytes total under a specific row. Useful for adding new rows to `_encoded`."""
+    def get_sum_of_bytes(self, until_row_index: int = -1) -> int:
+        """Get the total number of bytes that are accounted for.
+        This operation is O(1).
+
+        Args:
+            until_row_index (int): Optionally provide a row index for which the sum will end.
+                Gets all bytes accounted for up until `until_row_index`.
+
+        Returns:
+            int: Number of bytes encoded at and below `until_row_index`.
+        """
 
         if len(self._encoded) == 0:
             return 0
 
-        if row_index < 0:
-            row_index = len(self._encoded) + row_index
+        if until_row_index < 0:
+            until_row_index = len(self._encoded) + until_row_index
+
+        last_last_seen_index = 0
+        if until_row_index > 0:
+            last_last_seen_index = self._encoded[
+                until_row_index - 1, LAST_SEEN_INDEX_COLUMN
+            ]
 
-        row = self._encoded[row_index]
+        row = self._encoded[until_row_index]
+        sb = row[START_BYTE_COLUMN]
+        nb = row[NUM_BYTES_COLUMN]
+        delta = row[LAST_SEEN_INDEX_COLUMN] - last_last_seen_index
 
-        if row_index == 0:
-            previous_last_index = -1
-        else:
-            previous_last_index = self._encoded[row_index - 1, LAST_SEEN_INDEX_COLUMN]
+        if until_row_index == 0:
+            delta += 1
 
-        num_samples = row[LAST_SEEN_INDEX_COLUMN] - previous_last_index
-        num_bytes_for_entry = num_samples * row[NUM_BYTES_COLUMN]
-        return int(num_bytes_for_entry + row[START_BYTE_COLUMN])
+        return sb + (nb * delta)
 
     def _validate_incoming_item(self, num_bytes: int, _):
         if num_bytes < 0:
             raise ValueError(f"`num_bytes` must be >= 0. Got {num_bytes}.")
 
         super()._validate_incoming_item(num_bytes, _)
 
-    def _combine_condition(self, num_bytes: int) -> bool:
-        last_num_bytes = self._encoded[-1, NUM_BYTES_COLUMN]
+    def _combine_condition(self, num_bytes: int, compare_row_index: int = -1) -> bool:
+        """Checks if `num_bytes` matches the `num_bytes` represented at row with index `compare_row_index`."""
+
+        last_num_bytes = self._encoded[compare_row_index, NUM_BYTES_COLUMN]
         return num_bytes == last_num_bytes
 
-    def _make_decomposable(self, num_bytes: int) -> Sequence:
-        sb = self.num_bytes_encoded_under_row(-1)
+    def _make_decomposable(
+        self, num_bytes: int, compare_row_index: int = -1
+    ) -> Sequence:
+        sb = self.get_sum_of_bytes(compare_row_index)
         return [num_bytes, sb]
 
+    def _post_process_state(self, start_row_index: int):
```

Comment:
can we rename this to update_start_bytes or something similar?

Reply:
it's a general function, future encoders might override it. doesn't make sense to rename